### PR TITLE
BUG: isetitem coercing df columns to object

### LINKED ISSUE #49922: BUG: `frame.isetitem` incorrectly casts `Int64` columns to dtype `object`
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this bug exists on the main branch of pandas.

### Reproducible Example

```python
import pandas as pd

df1 = pd.DataFrame({"foo": [0, 0, 0]}, index=[1, 2, 3])
df2 = pd.DataFrame({"foo": [1, 1, 1]}, index=[2, 3, 4])

int_cols = [0, 1]
full_df1 = pd.concat([df1, df2], axis="columns")
full_df2 = full_df1.copy()  # to showcase different behaviour

# My current code: Works, but started giving me random FutureWarning as of v1.5
# (apparently happens when df1 and df2 have the same index initially, not relevant to this bug)
full_df1.iloc[:, int_cols] = full_df1.iloc[:, int_cols].astype("Int64")
print(full_df1.dtypes)
# foo    Int64
# foo    Int64
# dtype: object

# My attempt to future-proof my code (as I quite often have columns that are non-unique):
full_df2.isetitem(int_cols, full_df2.iloc[:, int_cols].astype("Int64"))
print(full_df2.dtypes)
# foo    object
# foo    object
# dtype: object
```

### Issue Description

After trying to follow the instructions in the `FutureWarning` referenced in the code above, I got into trouble with nullable integer columns ending up as `dtype=object`.

The warning in question:

> FutureWarning: In a future version, `df.iloc[:, i] = newvals` will attempt to set the values inplace instead of always setting a new array. To retain the old behavior, use either `df[df.columns[i]] = newvals` or, if columns are non-unique, `df.isetitem(i, newvals)`

### Expected Behavior

I would expect that `frame.isetitem(cols, ...)` would not cast `Int64` columns to `object`, but rather work exactly similar to how `frame.iloc[:, cols] = ...` works.

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : 8dab54d6573f7186ff0c3b6364d5e4dd635ff3e7
python           : 3.8.13.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 22.1.0
Version          : Darwin Kernel Version 22.1.0: Sun Oct  9 20:15:09 PDT 2022; root:xnu-8792.41.9~2/RELEASE_ARM64_T6000
machine          : arm64
processor        : arm
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : None.UTF-8

pandas           : 1.5.2
numpy            : 1.23.2
pytz             : 2022.2.1
dateutil         : 2.8.2
setuptools       : 65.4.1
pip              : 22.3.1
Cython           : 0.29.32
pytest           : 6.2.5
hypothesis       : None
sphinx           : 5.1.1
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : 1.1
pymysql          : None
psycopg2         : None
jinja2           : 3.1.2
IPython          : 8.4.0
pandas_datareader: None
bs4              : 4.11.1
bottleneck       : None
brotli           : None
fastparquet      : None
fsspec           : None
gcsfs            : None
matplotlib       : 3.5.3
numba            : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pyreadstat       : None
pyxlsb           : None
s3fs             : None
scipy            : None
snappy           : None
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
zstandard        : None
tzdata           : None

</details>

### PR DESCRIPTION
**Pull request by phofl**
- [x] closes #49922 (Replace xxxx with the GitHub issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/core/frame.py [hunk 1]
```diff
@@ -3822,6 +3822,15 @@ def isetitem(self, loc, value) -> None:
         In cases where `frame.columns` is unique, this is equivalent to
         `frame[frame.columns[i]] = value`.
         """
+        if isinstance(value, DataFrame):
+            if is_scalar(loc):
+                loc = [loc]
+
+            for i, idx in enumerate(loc):
+                arraylike = self._sanitize_column(value.iloc[:, i])
+                self._iset_item_mgr(idx, arraylike, inplace=False)
```
Comment (MarcoGorelli) [inline pandas/core/frame.py:3831]: in the tests you've added, `i` and `idx` are always the same - is it possible to add a test where `i` and `idx` differ?

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
generally looks good, just got a suggestion to improve testing

File: pandas/core/frame.py (hunk 1 above)
Comment (MarcoGorelli) [reply pandas/core/frame.py:3831]: as in, if I changed this to
```
                arraylike = self._sanitize_column(value.iloc[:, i])
                self._iset_item_mgr(i, arraylike, inplace=False)
```
then there should be a test which fails

Comment (phofl) [reply pandas/core/frame.py:3831]: Good catch, adjusted one of the tests

**Review by MarcoGorelli — APPROVED** [review]
Looks good to me! Over to @jbrockmendel

File: pandas/tests/frame/indexing/test_setitem.py [hunk 2]
```diff
@@ -728,6 +728,36 @@ def test_setitem_rhs_dataframe(self):
         df.isetitem(0, DataFrame({"a": [10, 11]}, index=[1, 2]))
         tm.assert_frame_equal(df, expected)
 
+    def test_isetitem_ea_df(self):
+        # GH#49922
+        df = DataFrame([[1, 2, 3], [4, 5, 6]])
+        rhs = DataFrame([[11, 12], [13, 14]], dtype="Int64")
+
+        df.isetitem([0, 1], rhs)
```
Comment (jbrockmendel) [inline pandas/tests/frame/indexing/test_setitem.py:736]: nitpick can we make a test file for isetitem.  otherwise lgtm

**Review by jbrockmendel — APPROVED** [review]
LGTM

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @phofl

**Pull request merged**